Add read and write timeouts

Choraden opened the pull request:
This patch makes the proxy more configurable when it comes to waiting for read/write.
Global timeout is now exchanged for
 - ReadTimeout - maximum duration for reading the entire request, including the body
 - ReadHeaderTimeout - maximum duration for reading request headers
 - WriteTimeout - maximum duration for writing the response

Choraden committed:
proxy: Add read and write timeouts

This patch makes the proxy more configurable when it comes to waiting for read/write.
Added timeouts:
 - ReadTimeout - maximum duration for reading the entire request, including the body
 - ReadHeaderTimeout - maximum duration for reading request headers
 - WriteTimeout - maximum duration for writing the response

Timeouts logic comes from http.Server from stdlib - https://github.com/golang/go/blob/ea14d1b6e1167159dfc8408073ef411c3cf1d7e0/src/net/http/server.go#L961

mmatczuk commented on proxy.go:
```diff
@@ -63,9 +63,33 @@ type Proxy struct {
 	// ErrorResponse specifies a custom error HTTP response to send when a proxying error occurs.
 	ErrorResponse func(req *http.Request, err error) *http.Response
 
+	// ReadTimeout is the maximum duration for reading the entire
```
For ascetics could we add whitespace for the parameters above to?

mmatczuk commented on proxy.go:
```diff
@@ -126,11 +149,6 @@ func (p *Proxy) SetDownstreamProxyFunc(f func(*http.Request) (*url.URL, error))
 }
 
-// SetTimeout sets the request timeout of the proxy.
-func (p *Proxy) SetTimeout(timeout time.Duration) {
```
Let's leave that for the compatibility it would set both read and write t/o.

Choraden committed:
proxy: Add ReadHeaderTimeout test

This test switch ReadTimeout and WriteTimeout off.
Then sets ReadHeaderTimeout and expects a connection to timeout on reading a header.

mmatczuk commented on proxy_test.go:
```diff
@@ -1462,3 +1463,57 @@ func TestRacyClose(t *testing.T) {
+			return true
+		}
+
+		switch err {
```
So what do we expect exactly?

Choraden replied:
I find it confusing that some times I would get io.UnexpectedEOF the other time "connection reset by peer", so I decided to check for all these errors

Choraden replied:
n/m, I found a workaround

mmatczuk commented on proxy.go:
```diff
@@ -290,7 +313,30 @@ var (
+	if d := p.readHeaderTimeout(); d > 0 {
+		hdrDeadline = t0.Add(d)
+	}
+	if p.ReadTimeout > 0 {
```
It looks nicer in the original version to me, why change it.

```
	if d := c.server.readHeaderTimeout(); d > 0 {
		hdrDeadline = t0.Add(d)
	}
	if d := c.server.ReadTimeout; d > 0 {
		wholeReqDeadline = t0.Add(d)
	}
```

mmatczuk commented on proxy.go:
```diff
@@ -309,6 +355,13 @@ func (p *Proxy) readRequest(ctx *Context, conn net.Conn, brw *bufio.ReadWriter)
 		}
 	}
 
+	// Adjust the read deadline if necessary.
```
Nice trick.

mmatczuk commented on proxy.go:
```diff
@@ -585,6 +638,12 @@ func (p *Proxy) handle(ctx *Context, conn net.Conn, brw *bufio.ReadWriter) error
 		}
 	}
 
+	if p.WriteTimeout != 0 {
```
Maybe `> 0` ?
